### ISSUE TITLE
Phase 4

### ISSUE THREAD

**Issue by brandy333-umbre** [brandy333-umbre/medical-no-show-appointments#3]
train_model.py:
from __future__ import annotations

import json
import os
import numpy as np
import pandas as pd
import joblib

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    roc_auc_score,
    average_precision_score,
    confusion_matrix,
    classification_report,
    precision_score,
    recall_score,
    f1_score,
)

DATA_PATH = "data/noshow.csv"
PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
ARTIFACTS_DIR = os.path.join(PROJECT_ROOT, "artifacts")

def load_and_clean(path: str) -> pd.DataFrame:
    df = pd.read_csv(path).copy()

    # Target: "Yes" = no-show (1), "No" = showed up (0)
    df["target"] = (df["No-show"] == "Yes").astype(int)

    # Parse datetimes
    df["ScheduledDay"] = pd.to_datetime(df["ScheduledDay"], errors="coerce", utc=True)
    df["AppointmentDay"] = pd.to_datetime(df["AppointmentDay"], errors="coerce", utc=True)

    # Feature engineering
    df["lead_time_hours"] = (df["AppointmentDay"] - df["ScheduledDay"]).dt.total_seconds() / 3600.0
    df["scheduled_hour"] = df["ScheduledDay"].dt.hour
    df["scheduled_weekday"] = df["ScheduledDay"].dt.day_name()
    df["appointment_weekday"] = df["AppointmentDay"].dt.day_name()

    # Fix known data issue
    df = df[df["Age"].between(0, 120)]

    # Drop identifiers / leakage-prone columns + raw target/datetimes
    drop_cols = ["PatientId", "AppointmentID", "No-show", "ScheduledDay", "AppointmentDay"]
    df = df.drop(columns=[c for c in drop_cols if c in df.columns], errors="ignore")

    return df

def build_model(X: pd.DataFrame) -> Pipeline:
    # Column types (include "string" to avoid pandas warning in newer versions)
    num_cols = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
    cat_cols = X.select_dtypes(include=["object", "string", "bool"]).columns.tolist()

    numeric_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler()),
        ]
    )

    categorical_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="most_frequent")),
            ("onehot", OneHotEncoder(handle_unknown="ignore")),
        ]
    )

    preprocessor = ColumnTransformer(
        transformers=[
            ("num", numeric_transformer, num_cols),
            ("cat", categorical_transformer, cat_cols),
        ],
        remainder="drop",
    )

    clf = LogisticRegression(
        max_iter=2000,
        class_weight="balanced",
        solver="liblinear",
    )

    return Pipeline(
        steps=[
            ("preprocess", preprocessor),
            ("clf", clf),
        ]
    )

def evaluate_at_threshold(y_true: np.ndarray, proba: np.ndarray, threshold: float) -> dict:
    preds = (proba >= threshold).astype(int)
    return {
        "threshold": threshold,
        "precision": precision_score(y_true, preds, zero_division=0),
        "recall": recall_score(y_true, preds, zero_division=0),
        "f1": f1_score(y_true, preds, zero_division=0),
    }

def tune_thresholds(
    y_true: np.ndarray,
    proba: np.ndarray,
    target_recall: float = 0.70,
    n_thresholds: int = 101,
) -> tuple[pd.DataFrame, float, float]:
    thresholds = np.linspace(0.0, 1.0, n_thresholds)
    rows = [evaluate_at_threshold(y_true, proba, t) for t in thresholds]
    df = pd.DataFrame(rows)

    # Best threshold by F1
    best_f1_row = df.loc[df["f1"].idxmax()]
    best_f1_threshold = float(best_f1_row["threshold"])

    # Threshold achieving at least target recall with best precision (or closest if none)
    df_meet = df[df["recall"] >= target_recall]
    if len(df_meet) > 0:
        best_recall_threshold = float(df_meet.sort_values(["precision", "threshold"], ascending=[False, True]).iloc[0]["threshold"])
    else:
        # If target recall is impossible, choose the closest recall
        best_recall_threshold = float(df.iloc[(df["recall"] - target_recall).abs().idxmin()]["threshold"])

    return df, best_f1_threshold, best_recall_threshold

def print_eval(y_true: np.ndarray, proba: np.ndarray, threshold: float, title: str) -> None:
    preds = (proba >= threshold).astype(int)
    cm = confusion_matrix(y_true, preds)
    print(f"\n=== {title} ===")
    print(f"Threshold: {threshold:.2f}")
    print("Confusion matrix (TN FP / FN TP):")
    print(cm)
    print("\nClassification report:")
    print(classification_report(y_true, preds, digits=4))

def save_artifacts(
    model: Pipeline,
    best_f1_threshold: float,
    best_recall_threshold: float,
    roc_auc: float,
    pr_auc: float,
    best_f1_row: pd.Series,
    best_recall_row: pd.Series,
) -> None:
    """Persist model, threshold, and training metrics to artifacts/."""
    os.makedirs(ARTIFACTS_DIR, exist_ok=True)

    # model.joblib
    model_path = os.path.join(ARTIFACTS_DIR, "model.joblib")
    joblib.dump(model, model_path)
    print(f"\nSaved model        → {model_path}")

    # threshold.json
    threshold_data = {
        "best_f1_threshold": best_f1_threshold,
        "best_recall_threshold": best_recall_threshold,
        "deployment_threshold": best_recall_threshold,
        "note": "deployment_threshold defaults to best_recall_threshold (recall >= 0.70)",
    }
    threshold_path = os.path.join(ARTIFACTS_DIR, "threshold.json")
    with open(threshold_path, "w") as f:
        json.dump(threshold_data, f, indent=2)
    print(f"Saved thresholds   → {threshold_path}")

    # training_metrics.json
    metrics_data = {
        "roc_auc": round(roc_auc, 4),
        "pr_auc": round(pr_auc, 4),
        "best_f1_threshold": {
            "threshold": round(best_f1_threshold, 4),
            "precision": round(float(best_f1_row["precision"]), 4),
            "recall": round(float(best_f1_row["recall"]), 4),
            "f1": round(float(best_f1_row["f1"]), 4),
        },
        "best_recall_threshold": {
            "threshold": round(best_recall_threshold, 4),
            "precision": round(float(best_recall_row["precision"]), 4),
            "recall": round(float(best_recall_row["recall"]), 4),
            "f1": round(float(best_recall_row["f1"]), 4),
        },
    }
    metrics_path = os.path.join(ARTIFACTS_DIR, "training_metrics.json")
    with open(metrics_path, "w") as f:
        json.dump(metrics_data, f, indent=2)
    print(f"Saved metrics      → {metrics_path}")

def main() -> None:
    data_path = os.path.join(PROJECT_ROOT, DATA_PATH)
    df = load_and_clean(data_path)
    X = df.drop(columns=["target"])
    y = df["target"].to_numpy()

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )

    model = build_model(X_train)
    model.fit(X_train, y_train)

    proba = model.predict_proba(X_test)[:, 1]

    roc_auc = roc_auc_score(y_test, proba)
    pr_auc = average_precision_score(y_test, proba)

    print("=== Evaluation (Threshold-independent) ===")
    print(f"ROC-AUC: {roc_auc:.4f}")
    print(f"PR-AUC:  {pr_auc:.4f}")

    # Tune thresholds
    metrics_df, best_f1_t, best_recall_t = tune_thresholds(y_test, proba, target_recall=0.70)

    # Show a quick summary of the best points
    best_f1_row = metrics_df.loc[metrics_df["f1"].idxmax()]
    print("\n=== Best threshold by F1 ===")
    print(best_f1_row.to_string(index=False))

    # Find the row used for target recall threshold (closest match)
    best_recall_row = metrics_df.iloc[(metrics_df["threshold"] - best_recall_t).abs().idxmin()]
    print("\n=== Threshold targeting recall >= 0.70 (best precision among those) ===")
    print(best_recall_row.to_string(index=False))

    # Print detailed confusion matrices / reports for both thresholds
    print_eval(y_test, proba, best_f1_t, "Evaluation at Best-F1 Threshold")
    print_eval(y_test, proba, best_recall_t, "Evaluation at Target-Recall Threshold (>= 0.70)")

    # Save all artifacts
    save_artifacts(model, best_f1_t, best_recall_t, roc_auc, pr_auc, best_f1_row, best_recall_row)

    print("\nTraining complete. Artifacts saved to artifacts/")

if __name__ == "__main__":
    main()

output:
=== Evaluation (Threshold-independent) ===
ROC-AUC: 0.6697
PR-AUC:  0.3067

=== Best threshold by F1 ===
0.470000
0.297806
0.665995
0.411573

=== Threshold targeting recall >= 0.70 (best precision among those) ===
0.450000
0.284779
0.721326
0.408344

=== Evaluation at Best-F1 Threshold ===
Threshold: 0.47
Confusion matrix (TN FP / FN TP):
[[10632  7010]
 [ 1491  2973]]

Classification report:
              precision    recall  f1-score   support

           0     0.8770    0.6027    0.7144     17642
           1     0.2978    0.6660    0.4116      4464

    accuracy                         0.6154     22106
   macro avg     0.5874    0.6343    0.5630     22106
weighted avg     0.7600    0.6154    0.6532     22106

=== Evaluation at Target-Recall Threshold (>= 0.70) ===
Threshold: 0.45
Confusion matrix (TN FP / FN TP):
[[9555 8087]
 [1244 3220]]

Classification report:
              precision    recall  f1-score   support

           0     0.8848    0.5416    0.6719     17642
           1     0.2848    0.7213    0.4083      4464

    accuracy                         0.5779     22106
   macro avg     0.5848    0.6315    0.5401     22106
weighted avg     0.7636    0.5779    0.6187     22106

Saved model        → c:\Users\brand\Desktop\medical-no-show-appointment\artifacts\model.joblib
Saved thresholds   → c:\Users\brand\Desktop\medical-no-show-appointment\artifacts\threshold.json
Saved metrics      → c:\Users\brand\Desktop\medical-no-show-appointment\artifacts\training_metrics.json

Training complete. Artifacts saved to artifacts/

predict.py output:
predict.py
c:\Users\brand\Desktop\medical-no-show-appointment\src\predict.py:149: DeprecationWarning:
        on_event is deprecated, use lifespan event handlers instead.

        Read more about it in the
        [FastAPI docs for Lifespan Events](https://fastapi.tiangolo.com/advanced/events/).

  @application.on_event("startup")
============================================================
PREDICTION DEMO
============================================================
Deployment threshold: 0.4500

--- Single predictions ---

Patient 1: Age=62, Gender=F, LeadTime=48.0h
  → no_show_probability: 0.3600
  → flag_for_intervention: False

Patient 2: Age=22, Gender=M, LeadTime=336.0h
  → no_show_probability: 0.5012
  → flag_for_intervention: True

Patient 3: Age=45, Gender=F, LeadTime=2.0h
  → no_show_probability: 0.4688
  → flag_for_intervention: True

--- Batch prediction ---

  Patient 1: probability=0.3600, intervene=False
  Patient 2: probability=0.5012, intervene=True
  Patient 3: probability=0.4688, intervene=True

interpretation: any patient whose predicted no-show probability is at or above 0.45 gets flagged for intervention (e.g., a reminder call or SMS).

monitor.py script output:
========================================================================
PRODUCTION MONITORING SIMULATION
========================================================================

Reference window:  88,420 records
Production window: 22,106 records
Deployment threshold: 0.4500

========================================================================
DATA DRIFT MONITORING REPORT
========================================================================

   Age                        [Kolmogorov-Smirnov]
     statistic: 0.0067   p-value: 0.398180   → OK
     ref: mean=37.0622, std=23.0970
     prod: mean=37.1973, std=23.1622

!! lead_time_hours            [Kolmogorov-Smirnov]
     statistic: 0.0338   p-value: 0.000000   → DRIFT DETECTED
     ref: mean=232.6830, std=365.8876
     prod: mean=234.9929, std=365.7456

!! appointment_weekday        [Chi-squared]
     statistic: 2606.3851   p-value: 0.000000   → DRIFT DETECTED

!! scheduled_weekday          [Chi-squared]
     statistic: 341.8724   p-value: 0.000000   → DRIFT DETECTED

!! no_show_rate               [Two-proportion Z-test]
     statistic: -5.2254   p-value: 0.000000   → DRIFT DETECTED
     ref rate: 0.2051   prod rate: 0.1893

------------------------------------------------------------------------
ACTION REQUIRED: Significant drift detected. Consider retraining.

========================================================================
PERFORMANCE MONITORING REPORT
========================================================================

  Deployment threshold: 0.4500
  Samples evaluated:   22,106
  Positive rate:       0.1893

  ROC-AUC:             0.6688  (training: 0.6697, delta: -0.0009)
  Precision:           0.2644
  Recall:              0.7331
  F1:                  0.3886

------------------------------------------------------------------------
Performance within acceptable bounds. No retraining needed.

========================================================================
FAIRNESS MONITORING REPORT
========================================================================

  --- Fairness by Neighbourhood (top groups) ---
  Group                           N  NoShow%   Recall     Prec       F1
  MARIA ORTIZ                  1216   17.52%   0.7324   0.2609   0.3847
  RESISTÊNCIA                  1037   17.84%   0.6595   0.2785   0.3917
  JARDIM CAMBURI                926   16.63%   0.3442   0.2000   0.2530
  JARDIM DA PENHA               864   15.86%   0.5912   0.1901   0.2877
  CENTRO                        682   20.38%   0.7842   0.2802   0.4129
  TABUAZEIRO                    663   21.12%   0.7214   0.3137   0.4372
  SANTO ANDRÉ                   640   15.00%   0.7812   0.2632   0.3937
  JESUS DE NAZARETH             639   22.85%   0.8425   0.3067   0.4497
  SANTA MARTHA                  629   18.12%   0.6316   0.3564   0.4557
  SANTO ANTÔNIO                 547   16.09%   0.6364   0.2679   0.3771

  Max recall gap across groups: 0.4983
  → Large disparity — investigate potential bias

  --- Fairness by Age Group (top groups) ---
  Group                           N  NoShow%   Recall     Prec       F1
  0-18                         5159   21.38%   0.8277   0.2721   0.4095
  19-35                        4670   22.76%   0.8260   0.3118   0.4527
  36-50                        4480   18.66%   0.7177   0.2676   0.3899
  51-65                        4427   15.54%   0.6003   0.2216   0.3237
  66+                          2710   14.87%   0.4467   0.2048   0.2808

  Max recall gap across groups: 0.3810
  → Large disparity — investigate potential bias

  --- Fairness by Scholarship (Socioeconomic Proxy) (top groups) ---
  Group                           N  NoShow%   Recall     Prec       F1
  No scholarship              19889   18.62%   0.7136   0.2629   0.3842
  Has scholarship              2217   21.70%   0.8836   0.2744   0.4187

  Max recall gap across groups: 0.1700
  → Moderate disparity — monitor closely

========================================================================
MONITORING COMPLETE
========================================================================

interpretation:
"Performance within acceptable bounds. No retraining needed." The model is stable.
